Mathlib/GroupTheory/GroupAction/Primitive.lean

```diff
@@ -0,0 +1,279 @@
+  Moreover, the classical notion is *primitive*,
+  which further assumes that `X` is not empty.
+
+- `MulAction.IsQuasiPreprimitive G X`
+  A structure that says that the action of the group `G` on the type `X` is *quasipreprimitive*,
+  namely, normal subgroups of `G` which act nontrivially act pretransitively.
+
+- We prove some straightforward theorems that relate preprimitivity
+  under equivariant maps, for images and preimages.
+
+## Relation with stabilizers
+
+- `MulAction.isSimpleOrderBlockMem_iff_isPreprimitive`
+  relates primitivity and the fact that the inclusion order on blocks containing is simple.
+
+- `MulAction.isCoatom_stabilizer_iff_preprimitive`
+  An action is preprimitive iff the stabilizers of points are maximal subgroups.
+
+- `MulAction.IsPreprimitive.isCoatom_stabilizer_of_isPreprimitive`
+  Stabilizers of points under a preprimitive action are maximal subgroups.
+
+## Relation with normal subgroups
+
+- `MulAction.IsPreprimitive.isQuasipreprimitive`
+  Preprimitive actions are quasipreprimitive.
+
+-/
+
+open Pointwise
+
+namespace MulAction
+
+variable (G : Type*) (X : Type*)
+
+-- Note : if the action is degenerate, singletons may not be blocks.
+/-- An additive action is preprimitive if it is pretransitive and
+the only blocks are the trivial ones -/
+class _root_.AddAction.IsPreprimitive [VAdd G X] extends AddAction.IsPretransitive G X : Prop where
+  /-- An action is preprimitive if it is pretransitive and
+  the only blocks are the trivial ones -/
+  isTrivialBlock_of_isBlock : ∀ {B : Set X}, AddAction.IsBlock G B → AddAction.IsTrivialBlock B
+
+/-- An action is preprimitive if it is pretransitive and
+the only blocks are the trivial ones -/
+@[to_additive]
+class IsPreprimitive [SMul G X] extends IsPretransitive G X : Prop where
+/-- An action is preprimitive if it is pretransitive and
+the only blocks are the trivial ones -/
+  isTrivialBlock_of_isBlock : ∀ {B : Set X}, IsBlock G B → IsTrivialBlock B
+
+open IsPreprimitive
+
+/-- An additive action of an additive group is quasipreprimitive if any normal subgroup
+that has no fixed point acts pretransitively -/
+class _root_.AddAction.IsQuasiPreprimitive
+    [AddGroup G] [AddAction G X] extends AddAction.IsPretransitive G X : Prop where
+  isPretransitive_of_normal :
+    ∀ {N : AddSubgroup G} [N.Normal], AddAction.fixedPoints N X ≠ .univ →
+      AddAction.IsPretransitive N X
+
+/-- An action of a group is quasipreprimitive if any normal subgroup
+that has no fixed point acts pretransitively -/
+@[to_additive]
+class IsQuasiPreprimitive [Group G] [MulAction G X] extends IsPretransitive G X : Prop where
+  isPretransitive_of_normal :
+    ∀ {N : Subgroup G} [N.Normal], fixedPoints N X ≠ .univ → IsPretransitive N X
+
+variable {G X}
+
+@[to_additive]
+theorem IsPreprimitive.subsingleton_or_eq_univ_of_isBlock
+    [SMul G X] [IsPreprimitive G X] {B : Set X} (hB : IsBlock G B) :
+    B.Subsingleton ∨ B = .univ :=
+  isTrivialBlock_of_isBlock hB
```

Comment:
I'm wondering. Should this be unnamespaced like so?
```suggestion
theorem subsingleton_or_eq_univ_of_isBlock
    [SMul G X] [IsPreprimitive G X] {B : Set X} (hB : IsBlock G B) :
    B.Subsingleton ∨ B = .univ :=
  isTrivialBlock_of_isBlock hB
```
There's no `IsPreprimitive` assumption to dot notate on and hopefully primitivity of an action will usually come from context, so no need to bother people with it in the name?

Reply:
Wouldn't IsBlock be an appropriate namespace?

Reply:
Ah yes, that was the extra argument: moving it out of the `IsPreprimitive` namespace means you can move it into the `IsBlock` namespace. So 👍 from me